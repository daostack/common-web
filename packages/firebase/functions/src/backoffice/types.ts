import { IBaseEntity } from "../util/types";


export interface ICircleBalanceBase extends IBaseEntity {
    id: string;
    account: string;
    available: [];
    unsettled: [];    
}

export interface ICircleBalancePayload {
    data: {
        available: any;
        unsettled: any;
    }
}

export type ICircleBalance = ICircleBalanceBase //ISubscriptionPayment | IProposalPayment;