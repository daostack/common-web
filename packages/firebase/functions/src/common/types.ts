import { IBaseEntity } from '../util/types';
import admin from 'firebase-admin';
import Timestamp = admin.firestore.Timestamp;

export interface ICommonEntity extends IBaseEntity {
  /**
   * The name of the common showed in the app and
   * other places (email, notification etc.)
   */
  name: string;

  /**
   * The URL of the image, used as header for
   * the common profile page
   */
  image: string;

  /**
   * The currently available funds of
   * the common in cents
   */
  balance: number;

  /**
   * The total amount of funds that the
   * common has raised to date in cents
   */
  raised: number;

  /**
   *  The timestamp after witch you are able to
   *  create funding proposals
   */
  fundingGoalDeadline: number;

  /**
   * List of all users, that are members of this common
   */
  members: ICommonMember[];

  /**
   * List of the rules, that a member must agree
   * to be a part if the common
   */
  rules: ICommonRule[];

  /**
   * List of links, that the common provided
   */
  links: ICommonLink[];

  /**
   * The common metadata properties
   */
  metadata: ICommonMetadata;

  /**
   * The whitelisting status of the common
   */
  register: CommonRegister;
}

export interface ICommonMember {
  userId: string;
  joinedAt: Timestamp;
}

export interface ICommonRule {
  /**
   * The title for the rule
   */
  title: string;

  /**
   * The description of the rule
   */
  value: string;
}

export interface ICommonLink {
  /**
   * The title of the link
   */
  title: string;

  /**
   * The address, to which the link is pointing
   */
  value: string;
}

export interface ICommonMetadata {
  byline: string;
  description: string;

  /**
   * The id of the user, who created the common
   */
  founderId: string;

  /**
   * The minimum amount in cents, required
   * to join the common
   */
  minFeeToJoin: number;

  /**
   * Whether the user should be charged every
   * month, that they are member of the common,
   * or only when they join
   */
  contributionType: ContributionType;
}

export interface ICommonUpdate {
  /**
   * The new common entity 
   */
  newCommon: ICommonEntity,

  /**
   * The userId of the user who is responsible for the change
   */
  changedBy: string
}

export type ContributionType = 'one-time' | 'monthly';

/**
 * Used to showcase whether the common is whitelisted
 *
 * na - The common is not whitelisted and thus visible only to members
 * registered - The common is whitelisted and part of the featured list
 */
export type CommonRegister = 'na' | 'registered';