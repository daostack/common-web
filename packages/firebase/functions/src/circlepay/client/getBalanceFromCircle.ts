import axios from 'axios';

import { externalRequestExecutor } from '../../util';
import { getCircleHeaders } from '../index';
import { circlePayApi } from '../../settings';
import { ErrorCodes } from '../../constants';


interface ICircleAmount {
  amount: string;
  currency: string;
}

interface ICircleBalance {
  data: {
    available: ICircleAmount[];
    unsettled: ICircleAmount[];
  }
}

export const getBalanceFromCircle = async (): Promise<ICircleBalance> => {
  const headers = await getCircleHeaders();

  return externalRequestExecutor<ICircleBalance>(async () => {
    return (await axios.get<ICircleBalance>(`${circlePayApi}/businessAccount/balances`, headers)).data;
  }, {
    errorCode: ErrorCodes.CirclePayError,
    message: 'Failed getting balance from circle'
  });
}