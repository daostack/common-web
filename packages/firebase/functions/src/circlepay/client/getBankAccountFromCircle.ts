import axios from 'axios';

import { externalRequestExecutor } from '../../util';
import { circlePayApi } from '../../settings';
import { ErrorCodes } from '../../constants';

import { ICircleGetBankAccountResponse } from '../cards/circleTypes';
import { getCircleHeaders } from '../index';

export const getBankAccountFromCircle = async (bankAccountId: string): Promise<ICircleGetBankAccountResponse> => {
  const headers = await getCircleHeaders();

  return  await externalRequestExecutor<ICircleGetBankAccountResponse>(async () => {
    return (await axios.get<ICircleGetBankAccountResponse>(`${circlePayApi}/banks/wires/${bankAccountId}`,
      headers
    )).data;
  }, {
    errorCode: ErrorCodes.CirclePayError,
    userMessage: 'Cannot get the bank account'
  });
}