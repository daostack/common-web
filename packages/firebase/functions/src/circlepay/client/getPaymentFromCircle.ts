import axios from 'axios';

import { externalRequestExecutor } from '../../util';
import { circlePayApi } from '../../settings';
import { ErrorCodes } from '../../constants';

import { getCircleHeaders } from '../index';
import { ICirclePayment } from '../types';

/**
 * Gets the current state of the payment from Circle
 *
 * @param paymentId - the Circle payment ID (not the local one)
 */
export const getPaymentFromCircle = async (paymentId: string): Promise<ICirclePayment> => {
  const headers = await getCircleHeaders();

  return externalRequestExecutor<ICirclePayment>(async () => {
    return (await axios.get<ICirclePayment>(`${circlePayApi}/payments/${paymentId}`, headers)).data;
  }, {
    errorCode: ErrorCodes.CirclePayError,
    message: `Circle call to GET payment with id ${paymentId} failed`
  });
};

