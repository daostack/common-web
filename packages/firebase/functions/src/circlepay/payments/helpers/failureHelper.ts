import { ICirclePayment } from '../../types';

import { IPaymentFailureReason, PaymentFailureResponseCodes } from '../types';

const failureCodeDescription: {
  [key in PaymentFailureResponseCodes]: string;
} = {
  card_account_ineligible: 'Ineligible account associated with card',
  card_invalid: 'Invalid card number',
  card_limit_violated: 'Exceeded amount or frequency limits',
  card_not_honored: 'Contact card issuer to query why payment failed',
  credit_card_not_allowed: 'Issuer did not support using a credit card for payment',
  payment_denied: 'Payment denied by Circle Risk Service or card processor risk controls',
  payment_failed: 'Payment failed due to unspecified error',
  payment_fraud_detected: 'Payment suspected of being associated with fraud',
  payment_not_funded: 'Insufficient funds in account to fund payment',
  payment_not_supported_by_issuer: 'Issuer did not support the payment',
  payment_stopped_by_issuer: 'A stop has been placed on the payment or card'

};

const getFailureDescription = (responseCode: PaymentFailureResponseCodes = 'payment_failed'): string =>
  failureCodeDescription[responseCode];

export const processFailedPayment = (payment: ICirclePayment): IPaymentFailureReason => {
  if (payment.data.status !== 'failed') {
    logger.warn('Trying to process failed payment that is not actually failed', {
      payment
    });

    return null;
  }

  const failureReason: IPaymentFailureReason = {
    errorCode: payment.data.errorCode || 'payment_failed',
    errorDescription: getFailureDescription(payment.data.errorCode)
  };

  logger.debug(`Processed failure reason for payment}`, {
    failureReasonFromCircle: payment.data.errorCode,
    processedFailureReason: failureReason,
    payment
  });

  return failureReason;
};

export const failureHelper = {
  processFailedPayment
};