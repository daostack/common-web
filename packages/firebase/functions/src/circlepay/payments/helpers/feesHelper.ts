import { ICirclePayment } from '../../types';
import { IPaymentFees } from '../types';

const defaultFees: IPaymentFees = {
  amount: 0,
  currency: 'USD'
};

export const processCircleFee = (circlePayment: ICirclePayment): IPaymentFees => {
  const { data: payment } = circlePayment;

  if (!payment.fees) {
    logger.warn('Trying to process payment fees, but there is no fees object on the payment', {
      payment
    });

    return defaultFees;
  }

  const fees: IPaymentFees = {
    // Circle returns the fees in dollars. Convert it to cents,
    amount: Number(payment.fees.amount) * 100,
    currency: payment.fees.currency
  };

  logger.debug('Converted circle fees', {
    circleFees: payment.fees,
    convertedFees: fees,
    circlePaymentId: circlePayment.data.id
  });

  return fees;
};

export const feesHelper = {
  processCircleFee
}