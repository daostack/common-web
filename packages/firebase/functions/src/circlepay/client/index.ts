import { getPaymentFromCircle } from './getPaymentFromCircle';
import { getCardFromCircle } from './getCardFromCircle';
import { getBankAccountFromCircle } from './getBankAccountFromCircle';

export { ICircleCard, CircleCardNetwork } from './getCardFromCircle';

export const circleClient = {
  getPayment: getPaymentFromCircle,
  getCard: getCardFromCircle,
  getBankAccount: getBankAccountFromCircle
}