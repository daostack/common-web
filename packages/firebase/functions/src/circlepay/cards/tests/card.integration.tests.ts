import { v4 } from 'uuid';
import firebaseFunctionsTests from 'firebase-functions-test';

import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

// Mockers
import '../../../util/tests/helpers/mockers/getUser.mocker';

import { circleApp, getTestAuthenticationToken } from '../../../util/tests/helpers';

import { ICircleCreateCardResponse } from '../../types';
import { IGetCircleCardResponse } from '../../client/getCardFromCircle';

import {
  createCardSuccessfulRequest,
  getCardFailedCvvCheck,
  getCardSuccessfulChecks
} from './data/testCardCirlceResponses';

import {
  invalidCreateCardRequest_MissingKeyId,
  invalidCreateCardRequest_PassedDate,
  validCreateCardRequest
} from './data/testCardRequests';

import { ErrorCodes } from '../../../constants';

const circleMatcher = /.+circle\.com\/.*/;

const test = firebaseFunctionsTests({
  projectId: 'common-tests'
});

describe('Card creation process', () => {
  afterAll(async () => {
    await test.cleanup();
  });

  it('should work', () => {
    expect(true).toBeTruthy();
  });

  it('should be healthy', async () => {
    const authToken = await getTestAuthenticationToken(v4());

    const response = await circleApp
      .get('/health')
      .set({
        Authorization: authToken
      });


    expect(response.status).toBe(200);
    expect(response.body.healthy).toBeTruthy();
  });

  describe('Card creation validation', () => {
    it('should create the card with all the data valid ', async () => {
      const cardOwnerId = v4();
      const mock = new MockAdapter(axios);

      mock.onPost(new RegExp(circleMatcher.source + (/\/cards/).source))
        .reply<ICircleCreateCardResponse>(200, createCardSuccessfulRequest);

      mock.onGet(new RegExp(circleMatcher.source + (/\/cards/).source))
        .reply<IGetCircleCardResponse>(200, getCardSuccessfulChecks);

      const response = await circleApp
        .post('/create-card')
        .send(validCreateCardRequest)
        .set({
          Authorization: getTestAuthenticationToken(cardOwnerId)
        });

      expect(response.status).toBe(200);
      expect(response.body.ownerId).toEqual(cardOwnerId);

      mock.restore();
    });

    it('should fail if the encryption key ID was not provided', async () => {
      const cardOwnerId = v4();

      const response = await circleApp
        .post('/create-card')
        .send(invalidCreateCardRequest_MissingKeyId)
        .set({
          Authorization: getTestAuthenticationToken(cardOwnerId)
        });

      expect(response.status).toBe(422);
      expect(response.body.data.errorCode).toBe(ErrorCodes.ValidationError);
      expect(response.body.data.detailedErrors).toMatchSnapshot();
    });

    it('should fail if the card expiry date is in the past ', async () => {
      const cardOwnerId = v4();

      const response = await circleApp
        .post('/create-card')
        .send(invalidCreateCardRequest_PassedDate)
        .set({
          Authorization: getTestAuthenticationToken(cardOwnerId)
        });

      expect(response.status).toBe(422);
      expect(response.body.data.errorCode).toBe(ErrorCodes.ValidationError);
      expect(response.body.data.detailedErrors).toMatchSnapshot();
    });

    it('should fail if no body is provided whatsoever', async () => {
      const cardOwnerId = v4();

      const response = await circleApp
        .post('/create-card')
        .send()
        .set({
          Authorization: getTestAuthenticationToken(cardOwnerId)
        });

      expect(response.status).toBe(422);
      expect(response.body.data.detailedErrors).toMatchSnapshot();
    });
  });

  describe('General card creation safeguard', () => {
    it('should fail if the user does not exist', async () => {
      const cardOwnerId = '404';
      const mock = new MockAdapter(axios);

      mock.onPost(new RegExp(circleMatcher.source + (/\/cards/).source))
        .reply<ICircleCreateCardResponse>(200, createCardSuccessfulRequest);

      const response = await circleApp
        .post('/create-card')
        .send(validCreateCardRequest)
        .set({
          Authorization: getTestAuthenticationToken(cardOwnerId)
        });

      expect(response.status).toEqual(404);
      expect(response.body.error.includes('Cannot find 404 with identifier user.userId'));

      mock.restore();
    });

    it('should fail if the CVV is incorrect', async () => {
      const cardOwnerId = v4();
      const mock = new MockAdapter(axios);

      mock.onPost(new RegExp(circleMatcher.source + (/\/cards/).source))
        .reply<ICircleCreateCardResponse>(200, createCardSuccessfulRequest);

      mock.onGet(new RegExp(circleMatcher.source + (/\/cards/).source))
        .reply<IGetCircleCardResponse>(200, getCardFailedCvvCheck);

      const response = await circleApp
        .post('/create-card')
        .send(validCreateCardRequest)
        .set({
          Authorization: getTestAuthenticationToken(cardOwnerId)
        });

      expect(response.status).toBe(500);
      expect(response.body.data.errorCode).toBe(ErrorCodes.CvvVerificationFail);

      mock.restore();
    });
  });
});