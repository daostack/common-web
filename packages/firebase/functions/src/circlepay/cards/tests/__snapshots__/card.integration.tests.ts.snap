// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card creation process Card creation validation should fail if no body is provided whatsoever 1`] = `
Array [
  Object {
    "field": "billingDetails.name",
    "message": "You must provide your fist and last name",
  },
  Object {
    "field": "billingDetails.name",
    "message": "billingDetails.name is a required field",
  },
  Object {
    "field": "billingDetails.city",
    "message": "billingDetails.city is a required field",
  },
  Object {
    "field": "billingDetails.country",
    "message": "billingDetails.country is a required field",
  },
  Object {
    "field": "billingDetails.line1",
    "message": "billingDetails.line1 is a required field",
  },
  Object {
    "field": "billingDetails.postalCode",
    "message": "billingDetails.postalCode is a required field",
  },
  Object {
    "field": "keyId",
    "message": "keyId is a required field",
  },
  Object {
    "field": "encryptedData",
    "message": "encryptedData is a required field",
  },
  Object {
    "field": "expMonth",
    "message": "expMonth is a required field",
  },
  Object {
    "field": "expYear",
    "message": "expYear is a required field",
  },
]
`;

exports[`Card creation process Card creation validation should fail if the card expiry date is in the past  1`] = `
Array [
  Object {
    "field": "expYear",
    "message": "expYear must be greater than or equal to 2021",
    "value": 2020,
  },
]
`;

exports[`Card creation process Card creation validation should fail if the encryption key ID was not provided 1`] = `
Array [
  Object {
    "field": "keyId",
    "message": "keyId is a required field",
  },
]
`;
